sqlmesh/core/macros.py

```diff
@@ -370,9 +390,30 @@ def add_one(evaluator: MacroEvaluator, column: exp.Literal) -> exp.Add:
     registry_name = "macros"
 
     def __call__(
-        self, func: t.Callable[..., DECORATOR_RETURN_TYPE]
+        self,
+        func: t.Callable[..., DECORATOR_RETURN_TYPE],
     ) -> t.Callable[..., DECORATOR_RETURN_TYPE]:
-        wrapper = super().__call__(func)
+        spec = inspect.getfullargspec(func)
+
+        @wraps(func)
+        def _typed_func(*args: t.Any, **kwargs: t.Any) -> DECORATOR_RETURN_TYPE:
+            """Coerce arguments where possible to the user-defined type annotations."""
+            evaluator: MacroEvaluator = args[0]
+            argmap = inspect.getcallargs(func, *args, **kwargs)
+            for param, value in argmap.items():
+                coercible_type = spec.annotations.get(param)
+                if not coercible_type:
+                    continue
+                if isinstance(coercible_type, str):
+                    continue
+                argmap[param] = evaluator._coerce(value, coercible_type)
+                if (ix := spec.args.index(param)) > -1:
+                    args = args[:ix] + (argmap[param],) + args[ix + 1 :]
```

Comment:
is this code necessary? can you just pass argmap into func?

Reply:
Maybe, just need to account for variadic arguments. So `**argmap` wont work unless I pop the variadic key, which i can do actually. So lets simplify yes.